pkg/workload/schemachange/generate.go

```diff
@@ -0,0 +1,223 @@
+	// parsable as SQL when executed.
+	// Example: `CREATE TABLE {TableName} ({Columns})`
+	Template string
+}
+
+type compiledCase struct {
+	Code         pgcode.Code
+	TemplateName string
+}
+
+// Generate is a helper to produce a psuedo-random case for a given
+// [tree.Statement]. Functions may return ErrCaseNotPossible to indicate that
+// they can't return an appropriate value. Closures may be used to track state
+// of generation. For example, a "Table" function may set a local variable so
+// that a "Columns" function knows which table to select columns from rather
+// than buffering all possible columns into memory.
+// Usage:
+//
+//	Generate[*tree.AlterType](og.params.rng, og.produceError(), []GenerationCase{
+//		{pgcode.SuccessfulCompletion, `ALTER TYPE {Type} ADD VALUE {Value}`},
+//		{pgcode.FileAlreadyExists, `ALTER TYPE "NeverExists" ADD VALUE "Irrelevant"`},
+//	}, template.FuncMap{
+//		"Type": func() *tree.Name {
+//			// Types that implement tree.NodeFormatter should be used to properly escape values.
+//			return PickOne(og.params.rng, existingTypes)
+//		},
+//		"Value": func() string {
+//			return "foo"
+//		},
+//	})
+func Generate[T tree.Statement](
+	rng *rand.Rand, produceError bool, cases []GenerationCase, funcs template.FuncMap,
+) (T, pgcode.Code, error) {
+	var zero T
+	name := reflect.TypeOf(zero).Name()
+
+	tpl := template.New(name).Delims("{", "}").Funcs(funcs)
+
+	// TODO(chrisseto): Consider caching the compiled templates.
+	var compiledCases []compiledCase
+	for i, gc := range cases {
+		tplName := fmt.Sprintf("Case %d - %s", i, gc.Code.String())
+
+		// If produceError is true, exclude successful cases from our selection.
+		if gc.Code == pgcode.SuccessfulCompletion && produceError {
+			continue
+		}
+
+		caseTpl := tpl.New(tplName)
+		if _, err := caseTpl.Parse(gc.Template); err != nil {
+			return zero, pgcode.Code{}, errors.NewAssertionErrorWithWrappedErrf(err, "failed to parse %q", tplName)
+		}
+
+		compiledCases = append(compiledCases, compiledCase{
+			Code:         gc.Code,
+			TemplateName: tplName,
+		})
+	}
+
+	// Randomize which case we'll select.
+	rng.Shuffle(len(compiledCases), func(i, j int) {
+		compiledCases[i], compiledCases[j] = compiledCases[j], compiledCases[i]
+	})
+
+	// Prioritize cases that will succeed but pushing them to the front of our
+	// slice. If no successful case is possible, we'll fallback to error cases.
+	slices.SortStableFunc(compiledCases, func(a, b compiledCase) bool {
+		aIsSuccess := a.Code == pgcode.SuccessfulCompletion
+		bIsSuccess := b.Code == pgcode.SuccessfulCompletion
+
+		return aIsSuccess && !bIsSuccess
+	})
+
+	for _, c := range compiledCases {
+		var raw strings.Builder
+		if err := tpl.ExecuteTemplate(&raw, c.TemplateName, nil); err != nil {
+			// If a case isn't possible, continue iterating until we find a valid
+			// case.
+			if errors.Is(err, ErrCaseNotPossible) {
+				continue
+			}
+			return zero, pgcode.Code{}, errors.WithStack(err)
+		}
```

Comment:
I didn't understand this block of code. So, we just get all the compilable generation cases (stored in `compiledCases`) and each generation case's template has been Parse'd. What does this line mean `tpl.ExecuteTemplate(&raw, c.TemplateName, nil)`? Does it mean try to execute the this generation case's template with `nil` data? What does it even mean? When would the return `err` be nil (and when would it be non-nil)?